rpc/shared/shared/src/main/scala/io/udash/rpc/UdashRPCFramework.scala

```diff
@@ -138,56 +143,72 @@ trait UdashRPCFramework extends GetterRPCFramework with ProcedureRPCFramework wi
+        var errorMsg: String = null
+        var tpe: String = null
+        while (obj.hasNext) {
+          val fi = obj.nextField()
+          fi.fieldName match {
+            case "response" =>
+              response = RawValueCodec.read(fi)
+            case "exception" =>
+              val wrapper = fi.readList()
+              exceptionName = wrapper.nextElement().readString()
+              exception = exceptionsRegistry.get(exceptionName).read(wrapper.nextElement())
+            case "callId" =>
+              callId = fi.readString()
+            case "cause" =>
+              cause = fi.readString()
+            case "errorMsg" =>
+              errorMsg = fi.readString()
+            case "type" =>
+              tpe = fi.readString()
+          }
         }
-      }
 
-      tpe match {
-        case "RPCResponseSuccess" =>
-          RPCResponseSuccess(response.asInstanceOf[RawValue], callId)
-        case "RPCResponseFailure" =>
-          RPCResponseFailure(cause, errorMsg, callId)
+        tpe match {
+          case "RPCResponseSuccess" =>
+            RPCResponseSuccess(response.asInstanceOf[RawValue], callId)
+          case "RPCResponseFailure" =>
+            RPCResponseFailure(cause, errorMsg, callId)
+          case "RPCResponseException" =>
+            RPCResponseException(exceptionName, exception.asInstanceOf[Throwable], callId)
+        }
       }
-    }
 
-    override def write(output: Output, value: RPCResponse): Unit = {
-      val obj = output.writeObject()
-      value match {
-        case RPCResponseSuccess(response, callId) =>
-          obj.writeField("type").writeString("RPCResponseSuccess")
-          RawValueCodec.write(obj.writeField("response"), response)
-          obj.writeField("callId").writeString(callId)
-        case RPCResponseFailure(cause, errorMsg, callId) =>
-          obj.writeField("type").writeString("RPCResponseFailure")
-          obj.writeField("cause").writeString(cause)
-          obj.writeField("errorMsg").writeString(errorMsg)
-          obj.writeField("callId").writeString(callId)
+      override def write(output: Output, value: RPCResponse): Unit = {
+        val obj = output.writeObject()
+        value match {
+          case RPCResponseSuccess(response, callId) =>
+            obj.writeField("type").writeString("RPCResponseSuccess")
+            RawValueCodec.write(obj.writeField("response"), response)
+            obj.writeField("callId").writeString(callId)
+          case RPCResponseFailure(cause, errorMsg, callId) =>
+            obj.writeField("type").writeString("RPCResponseFailure")
+            obj.writeField("cause").writeString(cause)
+            obj.writeField("errorMsg").writeString(errorMsg)
+            obj.writeField("callId").writeString(callId)
+          case RPCResponseException(name, exception, callId) =>
+            obj.writeField("type").writeString("RPCResponseException")
+            val exceptionField = obj.writeField("exception").writeList()
```

Comment:
Actually, in `scala-commons` 1.19.3+, I have introduced a requirement that `ObjectOutput` and `ObjectInput` implementations preserve order of fields. Thanks to that, you can use an object here instead of a list, which would probably be a little more natural. That's because one field can now be deserialized based on value of some other field, which is exactly your case.

For example, look at [current implementation of key-value pair serialization](https://github.com/AVSystem/scala-commons/blob/master/commons-shared/src/main/scala/com/avsystem/commons/serialization/GenCodec.scala#L344).